File: DesignPatternsInAutomatedTesting-Series/PatternsInAutomation.Tests/Advanced/NullObject/Base/BasePage.cs
using System;
using OpenQA.Selenium;
using OpenQA.Selenium.Support.UI;

namespace PatternsInAutomatedTests.Advanced.NullObject.Base
{
    public abstract class BasePage
    {
        protected IWebDriver driver;
        protected WebDriverWait driverWait;

        public BasePage(IWebDriver driver)
        {
            this.driver = driver;
            // wait 30 seconds.
            this.driverWait = new WebDriverWait(driver, new TimeSpan(0, 0, 30));
        }

        public virtual string Url { get; }

        public virtual void Open(string part = "")
        {
            if (string.IsNullOrEmpty(this.Url))
            {
                throw new ArgumentException("The main URL cannot be null or empty.");
            }
            this.driver.Navigate().GoToUrl(string.Concat(this.Url, part));
        }
    }
}

File: DesignPatternsInAutomatedTestingJava-Series/src/test/java/pageobjectadvanced/BasePage.java
package pageobjectadvanced;

public abstract class BasePage<ElementsТ extends BasePageElements, AssertsТ extends BasePageAsserts> {
    protected final String url;

    public BasePage(String url) {
        this.url = url;
    }

    public BasePage() {
        this.url = null;
    }

    protected abstract ElementsТ elements();

    public void navigate(String part) {
        Driver.getBrowser().navigate().to(url.concat(part));
    }

    public void navigate() {
        Driver.getBrowser().navigate().to(url);
    }

    protected abstract AssertsТ asserts();
}
